lib/puppet/property/list.rb

```diff
@@ -38,14 +38,34 @@ def dearrayify(array)
         array.sort.join(delimiter)
       end
 
-      def should
+      def members
         return nil unless @should
 
-        members = @should
         #inclusive means we are managing everything so if it isn't in should, its gone
-        members = add_should_with_current(members, retrieve) if ! inclusive?
+        if inclusive?
+          @should
+        else
+          add_should_with_current(@should, retrieve)
+        end
+      end
+
+      def should
+        tmp = members
+        dearrayify(tmp) if ! tmp.nil?
+      end
+
+      # Returns any values from the list that were returned by
+      # retrieve but are not set as being managed.
+      def is_but_shouldnt
+        # We know that if inclusive is not set, we won't have to
+        # remove any values from the system.
+        inclusive? ? Set.new(retrieve) - Set.new(@should) : Set.new()
+      end
 
-        dearrayify(members)
+      # Returns any values from the list that we are managing but that
+      # were not returned by retrieve
+      def should_but_isnt
+        Set.new(@should) - Set.new(retrieve)
```

Comment:
I haven't looked closely, but have we changed the semantics of the list property? I ask because it's used by several types.